pkg/proxy/index/index_controller.go

```diff
@@ -0,0 +1,294 @@
+	shardBaseURLs       map[string]string
+}
+
+// Start the controller. It does not really do anything, but to keep the shape of a normal
+// controller, we keep it.
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer func() {
+		c.shardInformersLock.Lock()
+		defer c.shardInformersLock.Unlock()
+		for _, stopCh := range c.shardClusterWorkspaceStopCh {
+			close(stopCh)
+		}
+	}()
+
+	klog.Infof("Starting %s controller", controllerName)
+	defer klog.Infof("Shutting %s controller", controllerName)
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) enqueueShard(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Enqueueing ClusterWorkspaceShard %q", key)
+
+	c.queue.Add(key)
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	shard, err := c.clusterWorkspaceShardLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.shardInformersLock.Lock()
+			defer c.shardInformersLock.Unlock()
+
+			delete(c.shardClusterWorkspaceInformers, shard.Name)
+			delete(c.shardClusterWorkspaceStopCh, shard.Name)
+
+			return nil
+		}
+		return err
+	}
+
+	c.shardInformersLock.Lock()
+	defer c.shardInformersLock.Unlock()
+
+	if _, found := c.shardClusterWorkspaceInformers[shard.Name]; !found {
+		client, err := c.clientGetter(shard)
+		if err != nil {
+			return err
+		}
+		informer := tenancyinformer.NewClusterWorkspaceInformer(client.Cluster(logicalcluster.Wildcard), clusterWorkspaceResyncPeriod, nil)
+		informer.AddEventHandler(c.clusterWorkspaceHandler)
+
+		stopCh := make(chan struct{})
+		c.shardClusterWorkspaceInformers[shard.Name] = informer
+		c.shardClusterWorkspaceStopCh[shard.Name] = stopCh
+
+		go informer.Run(stopCh)
+
+		// no need to wait. We only care about events and they arrive when they arrive.
+	}
+
+	return nil
+}
+
+func (c *Controller) Lookup(logicalCluster logicalcluster.Name) (string, bool) {
+	if logicalCluster == tenancyv1alpha1.RootCluster {
+		return c.rootHost, true
+	}
+
+	c.lock.RLock()
```

Comment:
if the `Lookup` method is on the hot path we could optimize and use `COW` technique.

Reply:
If we ever get to that one can use some bucketing of the index and only copy small parts.